python/lsst/daf/butler/sql/singleDatasetQueryBuilder.py

```diff
@@ -242,38 +224,31 @@ def _columns(selectable, names):
         # full set of link names for this DatasetType
         links = list(datasetType.dimensions.links())
 
-        # Starting point for both subqueries below: a join of dataset to
-        # dataset_collection
+        # Starting point is a join of dataset to dataset_collection
         subJoin = datasetTable.join(
             datasetCollectionTable,
             datasetTable.columns.dataset_id == datasetCollectionTable.columns.dataset_id
         )
         subWhere = and_(datasetTable.columns.dataset_type_name == datasetType.name,
                         datasetCollectionTable.columns.collection.in_(collections))
 
-        # CASE clause that transforms collection name to position in the given
-        # list of collections
-        collorder = case([
-            (datasetCollectionTable.columns.collection == coll, pos) for pos, coll in enumerate(collections)
-        ])
-
-        # first GROUP BY sub-query, find minimum `collorder` for each DataId
-        columns = [functions.min(collorder).label("collorder")] + _columns(datasetTable, links)
-        groupSubq = select(columns).select_from(subJoin).where(subWhere)
-        groupSubq = groupSubq.group_by(*links)
-        groupSubq = groupSubq.alias("sub1" + datasetType.name)
-
-        # next combined sub-query
-        columns = [collorder.label("collorder")] + _columns(datasetTable, ["dataset_id"] + links)
-        combined = select(columns).select_from(subJoin).where(subWhere)
-        combined = combined.alias("sub2" + datasetType.name)
-
-        # now join these two
-        joinsOn = [groupSubq.columns.collorder == combined.columns.collorder] + \
-                  [groupSubq.columns[colName] == combined.columns[colName] for colName in links]
-
-        return cls(registry, fromClause=combined.join(groupSubq, and_(*joinsOn)),
-                   datasetType=datasetType, selectableForDataset=combined, addResultColumns=addResultColumns)
+        # CASE clause that transforms collection name to prefixed name, plus
+        # MIN window function, plus SUBSTR
+        nDigits = 3
+        collFmt = f"{{:0{nDigits}d}}{{}}"
+        collCase = case({coll: collFmt.format(pos, coll) for pos, coll in enumerate(collections)},
+                        value=datasetCollectionTable.columns.collection)
+        collMin = functions.min(collCase).over(partition_by=_columns(datasetTable, links))
+        firstColl = func.substr(collMin, nDigits+1).label("first_collection")
+
+        columns = _columns(datasetTable, ["dataset_id"] + links) + \
+            _columns(datasetCollectionTable, ["collection"]) + [firstColl]
+
+        subq = select(columns).select_from(subJoin).where(subWhere)
+        whereClause = subq.columns.collection == subq.columns.first_collection
```

Comment:
I don't see any CTE here.  Is SQLAlchemy auto-detecting that it should use that syntax, or did you decide not to use it in the end?

Reply:
There is no CTE in code indeed, with how it is structured now I don't think we can use CTE as we pass FromCaluse and WhereClause separately to class constructor. And in this particular case there is not much benefit in using CTE vs. subquery, but I still wanted it mentioned in docstring because Chris says it's his favorite as it makes things easier to optimize in Oracle. If it's confusing it may be better not to mention it at all.

Reply:
Probably best to just summarize this discussion in the docs - we *could* use CTE, but would have to restructure the code, and have not (yet) done so.